File: 916. Word Subsets/main.go
package main

func counter(str string) map[rune]int {
	cnt := make(map[rune]int)

	for _, c := range str {
		cnt[c]++
	}

	return cnt
}

func wordSubsets(words1 []string, words2 []string) []string {
	count_2 := make(map[rune]int)

	for _, w := range words2 {
		for key, value := range counter(w) {
			count_2[key] = max(count_2[key], value)
		}
	}

	res := []string{}

	for _, w := range words1 {
		flag := true

		cnt := counter(w)

		for key, value := range count_2 {
			if cnt[key] < value {
				flag = false
				break
			}
		}

		if flag {
			res = append(res, w)
		}
	}

	return res
}

File: 916. Word Subsets/main.py
from collections import Counter, defaultdict
from typing import List


class Solution:
    def wordSubsets(self, words1: List[str], words2: List[str]) -> List[str]:
        count_2 = defaultdict(int)

        for w in words2:
            count_w = Counter(w)
            for c, cnt in count_w.items():
                count_2[c] = max(count_2[c], cnt)
        
        res = []

        for w in words1:
            count_w = Counter(w)
            flag = True

            for c, cnt in count_2.items():
                if count_w[c] < cnt:
                    flag = False
                    break
            
            if flag:
                res.append(w)

        return res
        